### PR TITLE
feat(common): add a LocationUpgradeService to mimic AngularJS functionality

### PR DESCRIPTION
**Pull request by jasonaden**
This PR contains a handful of features related to location services.

## Read `history.state`

Previously there wasn't a way to retrieve history.state from the Location service. The only time the framework exposed this value was in navigation events. This meant if you weren't using the Angular router, there wasn't a way to get access to this history.state value other than going directly to the DOM.

This PR adds an API to retrieve the value of history.state. This will be useful and needed to provide a backwards-compatible Location service that can emulate AngularJS's $location service since we will need to be able to read the state data in order to produce AngularJS location transition events.

This feature will additionally be useful to any application that wants to access state data through Angular rather than going directly to the DOM APIs.

## Read URL parts

Without this change, the framework doesn't surface URL parts such as hostname, protocol, and port. This makes it difficult to rebuild a complete URL. This change provides new APIs to read these values.

## Location Upgrade (`$location`)

This PR provides a replacement for `$location`. The new service is written in Angular, and can be consumed into existing applications by using the downgraded version
of the provider.

Prior to this addition, applications upgrading from AngularJS to Angular could get into a situation where AngularJS wanted to control the URL, and would often parse or serialize the URL in a different way than Angular. Additionally, AngularJS was alerted to URL changes only through the `$digest` cycle. This provided a buggy feedback loop
from Angular to AngularJS.

With this new `LocationUpgradeProvider`, the `$location` methods and events are provided in Angular, and use Angular APIs to make updates to the URL. Additionally, change
s to the URL made by other parts of the Angular framework (such as the Router) will be listened for and will cause events to fire in AngularJS, but will no longer attempt
 to update the URL (since it was already updated by the Angular framework).

This centralizes URL reads and writes to Angular and should help provide an easier path to upgrading AngularJS applications to Angular.

### CODE REVIEW

File: packages/common/test/location/location_spec.ts [hunk 1]
```diff
@@ -6,7 +6,10 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-import {Location} from '@angular/common';
+import {CommonModule, Location, LocationStrategy, PlatformLocation} from '@angular/common';
+import {PathLocationStrategy} from '@angular/common/src/common';
```
Comment (alxhub) [inline packages/common/test/location/location_spec.ts]: `@angular/common` instead of the deep import.

File: packages/common/upgrade/rollup.config.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+/**
```
Comment (alxhub) [inline packages/common/upgrade/rollup.config.js]: Is this file needed anymore, now that packages are built with Bazel? I don't think anything uses this.

File: packages/common/testing/src/mock_platform_location.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    const {pathname, search, state: parsedState, hash} = this.parseChanges(state, newUrl);
+    this.urlChanges.unshift({...this.urlChanges[0], pathname, search, state: parsedState});
+  }
+
+  forward(): void { throw new Error('Not implemented'); }
+
+  back(): void { this.urlChanges.shift(); }
+
+  // History API isn't available on server, therefore return undefined
+  getState(): unknown { return this.state; }
```
Comment (alxhub) [inline packages/common/testing/src/mock_platform_location.ts]: The comment here is misleading - this does seem to actually be implemented (as it should be in a mock).

File: packages/common/testing/src/mock_platform_location.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  }
+
+  get url(): string { return `${this.pathname}${this.search}${this.hash}`; }
+
+  private setHash(value: string, oldUrl: string) {
+    if (this.hash === value) {
+      // Don't fire events if the hash has not changed.
+      return;
+    }
+    (this as{hash: string}).hash = value;
```
Comment (alxhub) [inline packages/common/testing/src/mock_platform_location.ts]: This is pretty confusing to me. Are we overwriting the getter for `hash` with a field? If this is the intention, a comment would be really welcome here to explain what's going on.

Or, perhaps have a private `hashOverride` field, and change the `hash` getter to return that if it's defined, and otherwise go to `this.urlChanges[0]`.

File: packages/common/src/location/location.ts [hunk 5]
```diff
@@ -57,6 +57,7 @@ export class Location {
   _platformStrategy: LocationStrategy;
   /** @internal */
   _platformLocation: PlatformLocation;
+  private urlChangeListeners: any[] = [];
```
Comment (alxhub) [inline packages/common/src/location/location.ts]: `any`.... 😢

File: packages/common/testing/src/mock_platform_location.ts [hunk 6]
```diff
@@ -12,10 +12,10 @@ import {Subject} from 'rxjs';
 
 function parseUrl(urlStr: string, baseHref: string) {
   const verifyProtocol = /^((http[s]?|ftp):\/\/)/;
-  let serverBase = '';
+  let serverBase;
```
Comment (alxhub) [inline packages/common/testing/src/mock_platform_location.ts]: I would prefer explicitly declaring the type here.

File: packages/common/testing/tsconfig-build.json [hunk 7]
```diff
@@ -1,26 +1,27 @@
 {
   "extends": "../tsconfig-build.json",
-
   "compilerOptions": {
     "baseUrl": ".",
     "rootDir": "../",
     "paths": {
-      "@angular/core": ["../../../dist/packages/core"],
-      "@angular/common": ["../../../dist/packages/common"]
+      "@angular/core": [
```
Comment (alxhub) [inline packages/common/testing/tsconfig-build.json]: The changes to this file seem unnecessary.

File: packages/common/upgrade/src/params.ts [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+    search: string,
+    hash: string,
+    hostname: string,
+    port: string,
+    pathname: string
+  };
+}
+
+/**
+ * A `AngularJSUrlCodec` that uses logic from AngularJS to serialize and parse URLs
```
Comment (alxhub) [inline packages/common/upgrade/src/params.ts]: I would appreciate a link here into the AngularJS codebase where this logic is copied from.

File: packages/common/upgrade/BUILD.bazel [hunk 9]
```diff
@@ -15,6 +15,7 @@ ng_module(
     deps = [
         "//packages/common",
         "//packages/core",
+        "//packages/upgrade",
```
Comment (alxhub) [inline packages/common/upgrade/BUILD.bazel]: Is this needed for future or previous commits? I would squash this into the commit that actually requires the dependency.

File: packages/common/upgrade/src/$location.ts [hunk 10]
```diff
@@ -0,0 +1,693 @@
...
+/**
+ * Docs TBD.
+ *
+ * @publicApi
+ */
+export class LocationUpgradeService {
+  private initalizing = true;
+  private updateBrowser = false;
+  private $$absUrl: string = '';
+  private $$url: string = '';
```
Comment (alxhub) [inline packages/common/upgrade/src/$location.ts]: Given that we intentionally provide these `$$` fields because people read from them (and so technically they're here to be interface compatible with AngularJS), I would go ahead and make them public API. We want our public API guard to ensure they're not changed without consideration.

File: packages/common/upgrade/src/$location.ts [hunk 11]
```diff
@@ -0,0 +1,693 @@
...
+const DOUBLE_SLASH_REGEX = /^\s*[\\/]{2,}/;
+const IGNORE_URI_REGEXP = /^\s*(javascript|mailto):/i;
+const DEFAULT_PORTS: {[key: string]: number} = {
+  'http:': 80,
+  'https:': 443,
+  'ftp:': 21
+};
+
+/**
+ * Docs TBD.
```
Comment (alxhub) [inline packages/common/upgrade/src/$location.ts]: When you do write docs, having links to the AngularJS code being duplicated here would be ideal.

File: packages/common/upgrade/rollup.config.js (hunk 2 above)
Comment (jasonaden) [reply packages/common/upgrade/rollup.config.js]: Removed it.

File: packages/common/testing/src/mock_platform_location.ts (hunk 3 above)
Comment (jasonaden) [reply packages/common/testing/src/mock_platform_location.ts]: Correct. This was old. Fixed.

File: packages/common/testing/src/mock_platform_location.ts (hunk 4 above)
Comment (jasonaden) [reply packages/common/testing/src/mock_platform_location.ts]: Okay. I refactored this a bit since it actually doesn't look like it was firing at the correct time anyway. It's now fixed.

File: packages/common/src/location/location.ts (hunk 5 above)
Comment (jasonaden) [reply packages/common/src/location/location.ts]: Legacy while coding as I didn't know signature. Good catch! Put the type info in here.

File: packages/common/testing/src/mock_platform_location.ts (hunk 6 above)
Comment (jasonaden) [reply packages/common/testing/src/mock_platform_location.ts]: 👍

File: packages/common/upgrade/src/$location.ts (hunk 10 above)
Comment (jasonaden) [reply packages/common/upgrade/src/$location.ts]: I'll have this discussion with Igor to see what he thinks. I'm inclined to leave it out because if people are already using TypeScript from Definitely Typed, they will not have these and would have to do casting anyway. [See type definition here.](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/angular/index.d.ts#L985)

File: packages/common/upgrade/src/$location.ts (hunk 11 above)
Comment (jasonaden) [reply packages/common/upgrade/src/$location.ts]: Agreed.

File: packages/common/upgrade/src/params.ts (hunk 8 above)
Comment (jasonaden) [reply packages/common/upgrade/src/params.ts]: Added a number of links here for where the logic comes from. It's in a number of different places, so yeah, I think documenting that is a good idea!

File: packages/common/upgrade/BUILD.bazel (hunk 9 above)
Comment (jasonaden) [reply packages/common/upgrade/BUILD.bazel]: Hmm... okay. I intentionally pulled this out to reduce the size of the following commit. But I can squash this back in with the final changes that will come with public API review.

File: packages/common/testing/tsconfig-build.json (hunk 7 above)
Comment (jasonaden) [reply packages/common/testing/tsconfig-build.json]: This appears to be the result of running formatting.

**Comment by jasonaden** [comment]
[Global Presubmit](https://test.corp.google.com/OCL:244851799:BASE:244953959:1556064346544:a9e82d9c)

File: packages/common/upgrade/src/$location.ts (hunk 10 above)
Comment (alxhub) [reply packages/common/upgrade/src/$location.ts]: That makes sense. I was more thinking that we want to make sure we don't inadvertently change any of this API, but maybe having the documentation in place that the class mirrors AngularJS's interface is enough.

File: tools/public_api_guard/common/upgrade.d.ts [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    serverBaseHref?: string;
+    urlCodec?: typeof UrlCodec;
+    useHash?: boolean;
+}
+
+export declare class LocationUpgradeModule {
+    static config(config?: LocationUpgradeConfig): ModuleWithProviders<LocationUpgradeModule>;
+}
+
+export declare class LocationUpgradeProvider {
```
Comment (IgorMinar) [inline tools/public_api_guard/common/upgrade.d.ts]: `$locationShimProvider` & `$locationShim`

File: tools/public_api_guard/common/upgrade.d.ts [hunk 13]
```diff
@@ -0,0 +1,108 @@
+export declare const $locationProvider: Function;
```
Comment (IgorMinar) [inline tools/public_api_guard/common/upgrade.d.ts]: remove and also the peerDep from @angular/common to @angular/upgrade

File: tools/public_api_guard/common/upgrade.d.ts [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+        hash: string;
+        hostname: string;
+        port: string;
+        pathname: string;
+    };
+}
+
+export declare const LOCATION_UPGRADE_CONFIGURATION: InjectionToken<LocationUpgradeConfig>;
+
+export declare const LOCATION_UPGRADE_MODULE = "LOCATION_UPGRADE_MODULE";
```
Comment (IgorMinar) [inline tools/public_api_guard/common/upgrade.d.ts]: remove

File: tools/public_api_guard/common/upgrade.d.ts [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+export declare class LocationUpgradeProvider {
+    constructor(ngUpgrade: UpgradeModule, location: Location, platformLocation: PlatformLocation, urlCodec: UrlCodec, locationStrategy: LocationStrategy);
+    $get(): LocationUpgradeService;
+    hashPrefix(prefix?: string): void;
+    html5Mode(mode?: any): void;
+}
+
+export declare class LocationUpgradeService {
```
Comment (IgorMinar) [inline tools/public_api_guard/common/upgrade.d.ts]: rename

**Review by IgorMinar — APPROVED** [review]
let's rename some of the symbols and remove the $locationProvider and peerDep on @angular/upgrade. the rest lgtm from the api perspective.

File: tools/public_api_guard/common/upgrade.d.ts (hunk 12 above)
Comment (jasonaden) [reply tools/public_api_guard/common/upgrade.d.ts]: 👍

File: tools/public_api_guard/common/upgrade.d.ts (hunk 13 above)
Comment (jasonaden) [reply tools/public_api_guard/common/upgrade.d.ts]: Done. It is now a silent dependency so we can still get access to `$injector`

File: tools/public_api_guard/common/upgrade.d.ts (hunk 14 above)
Comment (jasonaden) [reply tools/public_api_guard/common/upgrade.d.ts]: 👍
